_posts/2018-01-30-quicli.md

```diff
@@ -0,0 +1,294 @@
+	(pass `-vvv` for "debug" level logging).
+	
+[failure]: https://docs.rs/failure
+
+## What I don't want quicli to be
+
+I want to use quicli to quickly build CLI apps
+– And I want to stop there!
+If you are no longer writing a small CLI tools
+(I'd say something like 500 lines of code is no longer small),
+your problems are most likely more complicated
+than what quicli gives you tools for.
+At that point,
+you should consider replacing your usage of (parts of) quicli
+with more sophisticated approaches.
+
+Let's contrast that with Garrett's [stdcli].
+It (re-)exports a lot of sophisticated tools!
+As Garrett writes:
+
+> stdcli's goal is basically to make creating a CLI in rust more like creating one in python from a user experience and ergonomics point of view.
+> This means:
+> batteries [are] included,
+> almost everything you need is already imported […]
+
+quicli also re-exports a bunch of stuff,
+and I agree that these re-exports make the experience of writing CLIs quickly much more smooth.
+There are some traits that are very useful, and deserve to re-exported!
+I want to be very deliberate in what I re-export, though,
+and ideally I want to provide abstractions instead of full-featured tools.
+Let me give some reasoning for this.
+
+For example:
+Having Read and Write in the prelude seems like a good idea, but AtomicBool?
+I've never used that one the last 4 years I've written Rust code.
+Let's take a step back and look at it from another perspective:
+I'm pretty sure even Read and Write are only used in a few specific use cases.
+I say: We should try to identify these use cases,
+and provide convenience functions for those!
+
+I didn't want to re-export `env_logger`;
+instead I changed the `main!` macro
+(that quicli provides to set give you nice error handling
+and the ability to use `?` in your "main" function)
+to initialize the logger automatically.
+Similarly, I added simple file read/write functions.
+
+Writing this, the following dawned on me:
+
+## I ship leaky abstractions
+
+I'd rather introduce some (simple and leaky) abstractions
+(that promise to only be useful 80% of the time)
+instead trying to give the user everything they need.
+I'm not delivering building blocks here,
+I deliver concrete tools.
+(This is hard for me -- I really like abstract and generic things!)
+
+Another example (which I haven't implemented) that works this way:
+A `regex_matches` function
+that automatically uses `lazy_static` and only covers the simplest case
+(maybe even return a very simple type
+instead of a powerful iterator over capture results).
+Why?
+Simple: Have you seen the regex docs?
+They are _wonderful._
+Sadly, for newcomers/forgetful people/drunk programmers/etc.
+they are also _wonderfully complex._
+
+So, instead of offering the user "everything,"
+I want to introduce some abstractions that are simple to use,
+and have simple but useful examples.
+Not to get stuck,
```

Comment:
Were you intending this to be part of the previous sentence? Otherwise I have no idea what you were trying to convey here with this sentence or what you mean by `Not to get stuck`

Reply:
Messed up the fix, should've been "To not get stuck"